### PR TITLE
Add cloning support

### LINKED ISSUE #27: Use the `COPYFILE_FICLONE` constant
For faster copy.

https://nodejs.org/api/fs.html#fs_fs_copyfile_src_dest_flags_callback

### PR DESCRIPTION
**Pull request by Artoria2e5**
This commit makes COPYFILE_FICLONE the default. A lot of tests may need to be changed...

Fixes #27

### CODE REVIEW

File: test/async.js [hunk 1]
```diff
@@ -38,20 +40,27 @@ test('reject an Error on missing `destination`', async t => {
...
 	fs.writeFileSync(t.context.source, buffer);
-	await cpFile(t.context.source, t.context.destination);
+	await cpFile(t.context.source, t.context.destination, {clone: false});
+	t.true(buffer.equals(fs.readFileSync(t.context.destination)));
+});
+
+test('clone a big file', async t => {
+	const buffer = crypto.randomBytes(THREE_HUNDRED_KILO);
+	fs.writeFileSync(t.context.source, buffer);
+	await cpFile(t.context.source, t.context.destination, {clone: true});
```
Comment (Artoria2e5) [inline test/async.js:63]: Can't make this `force` yet -- I can't figure out how to use different install scripts for different OS. Since `cmd` is super incompatible with `sh` I don't think I can get away with some `if` statements. Otherwise I would be able to make a known-fs disk image and stuff.

**Comment by Artoria2e5** [comment]
Eugh, `xo` is supposed to come up with its own `project`...

File: test/async.js (hunk 1 above)
Comment (Artoria2e5) [reply test/async.js:63]: Oh, travis on Windows uses `mingw` sh. Guess I am free to do whatever.

File: index.d.ts [hunk 2]
```diff
@@ -6,6 +6,13 @@ declare namespace cpFile {
 		@default true
 		*/
 		readonly overwrite?: boolean;
+
+		/**
+		Perform a fast copy-on-write clone (reflink) if possible.
+
```
Comment (sindresorhus) [inline index.d.ts:12]: Not clear from the docs here what `force` does.

Maybe also mention some use-cases for needing to set it to `false` or `'force'` as I cannot think of anything reason to not have this be just `true` all the time.

File: .travis.yml [hunk 3]
```diff
@@ -6,5 +6,7 @@ language: node_js
 node_js:
   - '12'
   - '10'
+before_install:
+  - if command -v mount grep df head tail cut &>/dev/null; then mount | grep "^$(df -Pk . | head -n 2 | tail -n 1 | cut -f 1 -d ' ') "; fi
```
Comment (sindresorhus) [inline .travis.yml:11]: Can you add a code comment that says what this does? For future readers of the code.

File: package.json [hunk 4]
```diff
@@ -54,6 +55,9 @@
 		"sinon": "^9.0.0",
 		"tsd": "^0.11.0",
 		"uuid": "^7.0.2",
-		"xo": "^0.27.2"
+		"xo": "^0.27.2",
```
Comment (sindresorhus) [inline package.json:58]: Try upgrading it to the latest version. That version has built-in TS support.

File: index.js [hunk 5]
```diff
@@ -82,21 +127,40 @@ const checkSourceIsFile = (stat, source) => {
...
+	let flags = 0;
+	if (!options.overwrite) {
+		flags |= fsConstants.COPYFILE_EXCL;
+	}
+
+	if (options.clone === true) {
+		flags |= fsConstants.COPYFILE_FICLONE;
+	} else if (options.clone === 'force') {
+		if (!Object.prototype.hasOwnProperty.call(fsConstants, 'COPYFILE_FICLONE_FORCE')) {
+			throw new CpFileError(`Node ${version} does not understand cloneFile`);
```
Comment (sindresorhus) [inline index.js:158]: The flags were added in Node 10 and this package targets that version, so I think this check is moot.

https://github.com/nodejs/node/commit/a16d88d9e9a6581e463082549823189aba25ca76

File: progress-emitter.js [hunk 6]
```diff
@@ -1,20 +1,33 @@
...
+		/** @type {string} */
 		this._sourcePath = sourcePath;
+		/** @type {string} */
 		this._destinationPath = destinationPath;
+		/** @type {number} */
+		this.size = Infinity;
 	}
 
 	get writtenBytes() {
+		// @ts-ignore
```
Comment (sindresorhus) [inline progress-emitter.js:25]:
```suggestion
		// @ts-expect-error
```

File: index.js [hunk 7]
```diff
@@ -1,16 +1,55 @@
...
+		fs.chmod(destination, stats.mode)
+	]);
+};
+
+/**
+ * @param {string} source
+ * @param {string} destination
+ * @param {import('.').Options} options
+ * @param {ProgressEmitter} progressEmitter
+ */
```
Comment (sindresorhus) [inline index.js:34]: I think all the JSDoc comments should have been added in a separate PR. Try to keep that in mind in the future. It makes it much easier to review without all the unrelated noise.

File: fs.js [hunk 8]
```diff
@@ -69,8 +103,32 @@ exports.makeDirSync = path => {
 	}
 };
 
+/**
+ * @param {import("fs").PathLike} source
+ * @param {import("fs").PathLike} destination
+ * @param {number?} [flags]
+ */
```
Comment (sindresorhus) [inline fs.js:110]: I prefer the doc style without so many noisy star characters. Would be great if you could use this style everywhere. Also single-quotes.

```suggestion
/**
@param {import('fs').PathLike} source
@param {import('fs').PathLike} destination
@param {number?} [flags]
*/
```

**Comment by sindresorhus** [comment]
Thanks for working on this 🙌🏻

File: index.d.ts (hunk 2 above)
Comment (Artoria2e5) [reply index.d.ts:12]: This is analogous to gnu cp's `--reflink=auto` vs `--reflink`. From a usability standpoint, it is usually preferable to have a fallback (hence `true`), but when someone comes onto a new platform it's nice to have a test for whether reflink is supported (hence `"force"`)

File: package.json (hunk 4 above)
Comment (Artoria2e5) [reply package.json:58]: The current version already does have that commit. And to quote a common saying, "it works on my machine"...

Could be travis CI caching an old version in the `node_modules` for some other dep though.

Comment (sindresorhus) [reply package.json:58]: Sure, but the latest XO version upgrades the TS parsers and such.

File: progress-emitter.js [hunk 9]
```diff
@@ -1,20 +1,33 @@
+/// <reference path="index.d.ts" />
 'use strict';
-const EventEmitter = require('events');
+const { EventEmitter } = require('events');
```
Comment (sindresorhus) [inline progress-emitter.js:3]: Don't do unrelated changes.

File: index.js [hunk 10]
```diff
@@ -39,39 +76,47 @@ const cpFileAsync = async (source, destination, options, progressEmitter) => {
...
+	// If we make it here, we should have no other errors. No need for a flag.
+	return updateStats(source, destination);
 };
 
+/**
+@param {string} sourcePath
+@param {string} destinationPath
+@param {import('.').Options} options
+@returns {Promise<void> & import('.').ProgressEmitter}
+ */
```
Comment (sindresorhus) [inline index.js:88]:
```suggestion
*/
```

Applies everywhere.

**Comment by sindresorhus** [comment]
@Artoria2e5 Bump

**Comment by sindresorhus** [comment]
@Artoria2e5 Bump
